#158113796 Setup Database and Refactor code

asamisellah opened the pull request:
### What does this PR do?
Refactor the code to use blueprints and sets up connection of the app to the database 

### Description of Task to be completed?
- Install Postgres database(db)Create
- Connect db to api
- Create user table and requests table

** In refactor **
- Create a model file
- Create class User
- Create class Request
- Create user blueprint

### How should this be manually tested?
1. clone the [repo](https://github.com/asamisellah/Maintenance-Tracker)
2. open the folder in your text editor or IDE
3. `git checkout -b new-branch`
4. `git fetch ch-env-setup-157930685`

*Setup*
Make sure virtual environment is installed. Open terminal and run the following commands
1. `cd maintenance-tracker`
2.  `virtualenv venv` to create virtual environment
3.  `source venv/bin/activate` to run environment
4. `pip install -r requirements.txt`

*Database*
Prerequisites: Should have Postgres installed
Open terminal and run
``` $ sudo -u postgres psql
    $ CREATE DATABASE mtracker;
    $ \q   
    $ python schema.py
    $ python run.py
```
### Any background context you want to provide?
None

### What are the relevant pivotal tracker stories?
[#158141028](https://www.pivotaltracker.com/story/show/158141028/comments/190387102)

asamisellah committed:
[Chore #157930716] Create blueprint for users

asamisellah committed:
[Chore #158113796] Create connection to Database

asamisellah committed:
[Chore #158113796] Setup Schema to create users and requests tables

asamisellah committed:
[Chore #158113796 #157930716] Refactor and Setup DB
- Refactor code to use blueprints
- Store data in table users

asamisellah committed:
[Feature #158113767] Create user and store in database
-Sign in user

asamisellah committed:
[Feature #158113767] Create user and store details in the database
- Sign in user
- Sign out user

asamisellah committed:
[Feature #158113767] Add Token to handle user authentication
- Use Flask JWT Extended

asamisellah committed:
[Feature #158113767] Intergrate JWT for user authentication

daisymacharia commented on app/__init__.py:
```diff
@@ -1,110 +1,121 @@
-requests = []
-users = []
-session = {}
+app.config['JWT_SECRET_KEY'] = 'super-secret'
```
Is this your secret key? it should be secret (as the name suggests 😆 ) so put in in your env file and source it from there.

asamisellah replied:
The key is secret now, for sure. Included it in an env file which is gitignored

daisymacharia commented on app/__init__.py:
```diff
@@ -1,110 +1,121 @@
-        "email": request.json.get('email'),
-        "password": request.json.get("password")
-    }
+    new_user = model.User(
```
why do you have to use the instance of User as `model.User`? I see you are importing it as User from the models, can't it just be accessed as User?

asamisellah replied:
Aaah..you're right. I've made the change

daisymacharia commented on app/__init__.py:
```diff
@@ -1,110 +1,121 @@
+        return jsonify({"message": "All Fields Required"}), 400
+    elif new_user.email is None:
+        return jsonify({"message": "All Fields Required"}), 400
+    elif new_user.email is None:
```
you are validating email twice, check `line 34`

daisymacharia commented on app/__init__.py:
```diff
@@ -1,110 +1,121 @@
+        str(uuid.uuid4),
+        request.json.get("username"),
+        request.json.get("email"),
+        request.json.get("password")
```
make sure you hash your passwords before saving them to the db you can do that after comparing your password and confirm password

daisymacharia commented on app/__init__.py:
```diff
@@ -1,110 +1,121 @@
+        request.json.get('category'),
+        request.json.get('area')
+    )
+    request_data.create_request()
```
make sure you validate this data before you save it in your db

asamisellah replied:
I've validated the data for create request and update request, including signup user

daisymacharia commented on app/model.py:
```diff
@@ -0,0 +1,74 @@
+    return requests
+
+
+def get_user_request(username):
```
this method is similar to `line 64` did you intend for this to be `get_request_by_id` ?

asamisellah committed:
[Chore #157930716] Modify users and requests tables
- Add roles and id field to users table
- Add id and user_id to request table
- Drop and re-create tables

asamisellah committed:
[Feature #158143411 #158113767] Log in user securely using JWT to access requests

asamisellah committed:
[Feature #158143791 #158143726] Update a user request and store changes in database

asamisellah committed:
[Feature #158233981] Delete user request endpoint removes existing records from database

asamisellah committed:
[Feature #158114180] Add admin feature for getting all requests in the database

asamisellah committed:
[Chore #158141028] Implement feedback

daisymacharia commented on app/__init__.py:
```diff
@@ -1,169 +1,197 @@
+                       request.json.get("category").lower(),
+                       request.json.get("area").lower()
+                       )
+        return jsonify({"Success": user_request}), 200
```
remember the specifications are that a user should not be able to edit a request that has already been approved by the admin.

daisymacharia commented on app/model.py:
```diff
@@ -0,0 +1,114 @@
+    return user
+
+
+class UserRequest():
```
you are required to add a field for the request status that is only editable by the admin.
